src/core.rs

```diff
@@ -92,35 +93,32 @@ impl Port {
     /// [HugrError::InvalidPortDirection]
     #[inline]
     pub fn as_incoming(&self) -> Result<IncomingPort, HugrError> {
-        match self.direction() {
-            Direction::Incoming => Ok(IncomingPort {
-                index: self.index() as u16,
-            }),
-            dir @ Direction::Outgoing => Err(HugrError::InvalidPortDirection(dir)),
-        }
+        self.as_directed()
+            .left()
+            .ok_or(HugrError::InvalidPortDirection(self.direction()))
     }
 
     /// Converts to an [OutgoingPort] if this port is one; else fails with
     /// [HugrError::InvalidPortDirection]
     #[inline]
     pub fn as_outgoing(&self) -> Result<OutgoingPort, HugrError> {
+        self.as_directed()
+            .right()
+            .ok_or(HugrError::InvalidPortDirection(self.direction()))
+    }
+
+    /// Converts to either an [IncomingPort] or an [OutgoingPort], as appropriate.
+    #[inline]
+    pub fn as_directed(&self) -> Either<IncomingPort, OutgoingPort> {
```

Comment:
Possibly `in_or_out`? But happy with `as_directed` too

Reply:
I'd say `in_or_out` would be more similar to `direction`, where `as_*` is a cast.